src/if_statements.py

```diff
@@ -355,228 +343,225 @@ def emit_goto_or_early_return(context: Context, target: Node, body: Body) -> Non
+                    node = node.fallthrough_edge
+                    break
+                elif node.fallthrough_edge is if_node:
+                    # This node represents `if (cond) { node.conditional_edge } else { if_node }`.
+                    # Here, we recursively expand `if_node`, which follows the following pattern:
+                    # ```
+                    #   if (or_terms || !cond) {
+                    #       if (tail_cond) {            // tail_cond starts with the `if_node` condition
+                    #           next_node;              // the returned `node` from traverse(...)
+                    #       } else {
+                    #           node.conditional_edge
+                    #       }
+                    #   } else {
+                    #       node.conditional_edge
+                    #   }
+                    # ```
+                    # ...which can be re-written into a single if statement with `||`.
+                    # ```
+                    #   if (!(or_terms || !cond) || !tail_cond) {
+                    #       node.conditional_edge;
+                    #   } else {
+                    #       next_node;
+                    #   }
+                    # ```
+                    assert if_node in chained_cond_nodes
+                    node, tail_cond = traverse(
+                        if_node, else_node, node.conditional_edge
                     )
+                    or_terms = [
+                        join_conditions(or_terms + [cond.negated()], "||").negated(),
+                        tail_cond.negated(),
+                    ]
+                    break
+                else:
+                    # Both branches from the node point to other conditions, rather than the if/else nodes.
+                    # Like in the previous case, we'll need to recurse & create a parenthetical expression.
+                    # Although `node` doesn't point to `if_node`, the previous nodes whose conditions are
+                    # accumulated in `or_terms` do point there. `node` expands in the following pattern:
+                    # ```
+                    #   if (or_terms) {
+                    #       if_node;
+                    #   } else {
+                    #       if (tail_cond) {            // `tail_cond` starts with the `node` condition
+                    #           next_node;              // the returned `node` from traverse(...)
+                    #       } else {
+                    #           if_node;
+                    #       }
+                    #   }
+                    # ```
+                    # ...which can be re-written into a single if statement with `||`.
+                    # ```
+                    #   if (or_terms || !tail_cond) {
+                    #       if_node;
+                    #   } else {
+                    #       next_node;
+                    #   }
+                    # ```
```

Comment:
This is a very nice comment! Unfortunately I don't really get it because I don't understand what `traverse` recursively does, e.g. why we pass in conditional_edge to it and why the if statement ends up being `if (tail_cond) { next_node; } else { if_node; }`...